### PR TITLE
FIX: Quoting local dates bbcode regeneration

### PR DESCRIPTION
**Pull request by martin-brennan**
This commit allows quoting of discourse-local-date elements
and converts the quoted tags back into bbcode so that the
rendered quote will also render the discourse-local-date HTML.
This works on single dates as well as date ranges, and supports
all of the options used by discourse-local-date.

This also necessitated adding addTextDecorateCallback to the
to-markdown core lib (similar to addBlockDecorateCallback and
addTagDecorateCallback) to transform the text nodes between
date ranges to remove the -> in the final quote.

c.f. https://meta.discourse.org/t/quotes-that-contain-date-time/101999

chrome-capture-2022-5-20

### CODE REVIEW

File: app/assets/javascripts/discourse/app/lib/to-markdown.js [hunk 1]
```diff
@@ -48,6 +51,30 @@ export function clearBlockDecorateCallbacks() {
...
 }
 
+/**
+ * Allows to add support for custom text node transformations
+ * based on the next/previous elements.
+ *
+ * ```
+ * addTextDecorateCallback(function (text, nextElement, previousElement) {
+ *   if (
+ *     startRangeOpts &&
```
Comment (tgxworld) [inline app/assets/javascripts/discourse/app/lib/to-markdown.js:61]: If I'm just looking at the example, `startRangeOpts` is not defined anywhere which may be confusing.

File: plugins/discourse-local-dates/assets/javascripts/initializers/discourse-local-dates.js [hunk 2]
```diff
@@ -128,6 +151,53 @@ function initializeDiscourseLocalDates(api) {
       },
     },
   });
+
+  let startRangeOpts = null;
```
Comment (tgxworld) [inline plugins/discourse-local-dates/assets/javascripts/initializers/discourse-local-dates.js:155]: This feels like a "global" being used by two callback functions. I wonder if we can pass this between the callback functions instead with some kind of metadata object.

File: app/assets/javascripts/discourse/app/lib/to-markdown.js [hunk 3]
```diff
@@ -48,6 +51,30 @@ export function clearBlockDecorateCallbacks() {
...
+ * ```
+ */
+export function addTextDecorateCallback(callback) {
+  textDecorateCallbacks.push(callback);
+}
+
+export function clearTextDecorateCallbacks() {
+  textDecorateCallbacks = [];
+}
+
```
Comment (SamSaffron) [inline app/assets/javascripts/discourse/app/lib/to-markdown.js:77]: Does this belong in the plugin js api, at least so we call out it is an available extensibility point?

File: app/assets/javascripts/discourse/app/lib/to-markdown.js (hunk 1 above)
Comment (martin-brennan) [reply app/assets/javascripts/discourse/app/lib/to-markdown.js:61]: Oh yeah you are right, will simplify the example

File: app/assets/javascripts/discourse/app/lib/to-markdown.js (hunk 3 above)
Comment (martin-brennan) [reply app/assets/javascripts/discourse/app/lib/to-markdown.js:77]: Yes that would be better, I can move all three of these there and possibly name them a little differently, I think they are only used by one of our plugins ATM so I can change it there easily

File: plugins/discourse-local-dates/assets/javascripts/initializers/discourse-local-dates.js (hunk 2 above)
Comment (martin-brennan) [reply plugins/discourse-local-dates/assets/javascripts/initializers/discourse-local-dates.js:155]: Yeah I don't like it a whole lot, maybe I can see if I can pass through some sort of metadata object in all of these `to-markdown` callbacks? Something that is kept in memory through the whole `toMarkdown()` chain for one block of HTML defined here https://github.com/discourse/discourse/blob/7120a33398998a07359330bc1fd703096585ed87/app/assets/javascripts/discourse/app/lib/to-markdown.js#L866 ? 

So then you would get something like

```
const metadata = {};
for (const callback of blockDecorateCallbacks) {
  const result = callback.call(this, metadata, text);

  if (typeof result !== "undefined") {
    text = result;
  }
}
```

I am not sure if I will be able to make this data flow through the whole way though....maybe I can attach arbitrary metadata to the `next/previous` Element instead?

**Comment by martin-brennan** [comment]
@SamSaffron / @tgxworld thoughts on what Sam suggested here:

https://github.com/discourse/discourse/pull/17141#discussion_r901261359

In CvX's PR @tgxworld you said you would prefer not to make these functions public as part of the plugin API https://github.com/discourse/discourse/pull/16943 , is this still the case? If so I will not move to the public API. I only noticed this because the plugin-api changelog is wrong, since it was not updated in that PR when those functions were moved out of the plugin-api.js file

https://github.com/discourse/discourse/blob/7120a33398998a07359330bc1fd703096585ed87/docs/CHANGELOG-JAVASCRIPT-PLUGIN-API.md#L10-L14

File: plugins/discourse-local-dates/assets/javascripts/initializers/discourse-local-dates.js (hunk 2 above)
Comment (martin-brennan) [reply plugins/discourse-local-dates/assets/javascripts/initializers/discourse-local-dates.js]: Was able to do this by passing through a `metadata` object through the flows

**Comment by martin-brennan** [comment]
@tgxworld made your changes and also have another question

**Comment by tgxworld** [comment]
Personally, I still don't think this needs to be a public API until we have more request for this. Especially in this case where the "API" is for a core plugin.

**Comment by martin-brennan** [comment]
@tgxworld That's fine I will leave it where it is then, I've just updated the changelog in the PR because the 1.3.0 description was incorrect based on the actual changes in 711cd7c85d2e58e101fd82800954cbc94baa9003

**Comment by SamSaffron** [comment]
> Personally, I still don't think this needs to be a public API until we have more request for this. Especially in this case where the "API" is for a core plugin.

I certainly don't want to block stuff here. It is a tricky problem, maybe this entire thing should not even be a plugin? Anyway. let's just go ahead and merge this creating a public API is not a blocking thing here.

**Review by SamSaffron — APPROVED** [review]

**Pull request merged**